pkg/daemon/update.go

```diff
@@ -93,9 +103,63 @@ func getNodeRef(node *corev1.Node) *corev1.ObjectReference {
 	}
 }
 
-// finalizeAndReboot is the last step in an update(), and it can also
-// be called as a special case for the "bootstrap pivot".
-func (dn *Daemon) finalizeAndReboot(newConfig *mcfgv1.MachineConfig) (retErr error) {
+func reloadService(name string) error {
+	_, err := runGetOut("systemctl", "reload", name)
+	return err
+}
+
+// performPostConfigChangeAction takes action based on what postConfigChangeAction has been asked.
+// For non-reboot action, it applies configuration, updates node's config and state.
+// In the end uncordon node to schedule workload.
+// If at any point an error occurs, we reboot the node so that node has correct configuration.
+func (dn *Daemon) performPostConfigChangeAction(postConfigChangeActions []string, configName string) error {
+	if ctrlcommon.InSlice(postConfigChangeActionReboot, postConfigChangeActions) {
+		dn.logSystem("Rebooting node")
+		return dn.reboot(fmt.Sprintf("Node will reboot into config %s", configName))
+	}
+	if ctrlcommon.InSlice(postConfigChangeActionNone, postConfigChangeActions) {
+		if dn.recorder != nil {
+			dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeNormal, "Reboot", "Config changes do not require reboot.")
+		}
+		dn.logSystem("Node has Desired Config %s, skipping reboot", configName)
+	}
+	if ctrlcommon.InSlice(postConfigChangeActionReloadCrio, postConfigChangeActions) {
+		serviceName := "crio"
+		if err := reloadService(serviceName); err != nil {
+			dn.logSystem("Reloading %s configuration failed, node will reboot instead. Error: %v", serviceName, err)
```

Comment:
event here is what i mean @yuqi-zhang :laughing:

Reply:
added warning event

Reply:
perf!